### PR TITLE
Analyzer: Added binding analyzer, updated entity interface content analyzer, fixed argument exception

### LINKED ISSUE #1704: Signal Durable Entity w/Multiple Interfaces Issues Analyzer Warning
The following merged PR allows an entity to implement multiple interfaces (which is fantastic), however, the roslyn analyzer is reporting the following warning: "Entity interface contains members other than methods" when I call `SignalEntityAsync<>` from an activity function or http trigger in my example.

https://github.com/Azure/azure-functions-durable-extension/pull/1277/commits

Below is an example:

```
public interface ISubscribableStore
{
    void Subscribe(EntityId entityId);

    void Unsubscribe(EntityId entityId);
}

public interface IVehicleStore : ISubscribableStore
{
    void Update(Vehicle vehicle);
}

public async Task ExecuteAsync(
    [ActivityTrigger] IDurableActivityContext context, [DurableClient] IDurableEntityClient client)
{
    var request = context.GetInput<Request>();

    await client.SignalEntityAsync<IVehicleStore>(request.EntityKey, e => e.Update(request.Vehicle));
}
```

Thanks and let me know if there is any other information you may need!

### LINKED ISSUE #1879: Durable Functions analyzer should flag use of bindings in orchestrator functions
**Is your feature request related to a problem? Please describe.**
Customers may try to use input or output bindings in their orchestration code, which can lead to stuck orchestrations or runtime failures. Here is a simple example based on a recent case I investigated:

```csharp
[FunctionName("Foo")]
public static async Task<bool> RunOrchestrator(
    [OrchestrationTrigger] IDurableOrchestrationContext context,
    [DurableClient] IDurableClient starter)
{
    DurableOrchestrationStatus orchestrationStatus = await starter.GetStatusAsync(context.InstanceId);
    return orchestrationStatus != null;
}
```

The problem is that the `IDurableClient.GetStatusAsync` method is non-deterministic. All functions bindings except for `[OrchestrationTrigger]` should be considered non-deterministic.

**Are there any existing GitHub discussions or issues filed that help give some context to this proposal?**
Not that I'm aware of.

**Describe the solution you'd like**
We should emit a warning if a developer tries to include a binding in their orchestrator function.

**Describe alternatives you've considered**
I considered runtime checks, but I'm not sure if it's possible for a trigger to be able to know whether the target function is using other bindings. It's something we could look into, though.

We could also consider improving the analyzer to flag async code that doesn't come from `IDurableOrchestrationContext`. I know we tried to do this at one point, but I think we eventually decided against it? I unfortunately don't remember the exact reasoning.

**Additional context**
N/A

### LINKED ISSUE #1883: Analyzer 'Microsoft.Azure.WebJobs.Extensions.DurableTask.Analyzers.DeterministicMethodAnalyzer' threw an exception of type 'System.ArgumentException' with message 'SyntaxTree is not part of the compilation
### Description
After a build of my .NETCore 3.1 Function App using `Microsoft.Azure.WebJobs.Extensions.DurableTask` v2.4.3 with the (tranient dependency) `Microsoft.Azure.WebJobs.Extensions.DurableTask.Analyzers` v0.4.0 I see an error in the build output `Analyzer 'Microsoft.Azure.WebJobs.Extensions.DurableTask.Analyzers.DeterministicMethodAnalyzer' threw an exception of type 'System.ArgumentException' with message 'SyntaxTree is not part of the compilation`

### Expected behavior
No exceptions cause by running the analyzer, only those that are triggered by the analyzer

### Actual behavior
```
Severity	Code	Description	Project	File	Line	Suppression State
Error	AD0001	Analyzer 'Microsoft.Azure.WebJobs.Extensions.DurableTask.Analyzers.DeterministicMethodAnalyzer' threw an exception of type 'System.ArgumentException' with message 'SyntaxTree is not part of the compilation
Parameter name: syntaxTree'.
Exception occurred with following context:
Compilation: <project name>

System.ArgumentException: SyntaxTree is not part of the compilation
Parameter name: syntaxTree
   at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.GetSemanticModel(SyntaxTree syntaxTree, Boolean ignoreAccessibility)
   at Microsoft.Azure.WebJobs.Extensions.DurableTask.Analyzers.SyntaxNodeUtils.TryGetSemanticModelForSyntaxTree(SemanticModel model, SyntaxNode node, SemanticModel& newModel)
   at Microsoft.Azure.WebJobs.Extensions.DurableTask.Analyzers.SyntaxNodeUtils.TryGetITypeSymbol(SemanticModel semanticModel, SyntaxNode node, ITypeSymbol& typeSymbol)
   at Microsoft.Azure.WebJobs.Extensions.DurableTask.Analyzers.IOTypesAnalyzer.RegisterDiagnostic(CompilationAnalysisContext context, SemanticModel semanticModel, SyntaxNode method)
   at Microsoft.Azure.WebJobs.Extensions.DurableTask.Analyzers.DeterministicMethodAnalyzer.RegisterAnalyzers(CompilationAnalysisContext context)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.<>c.<ExecuteCompilationActionsCore>b__48_1(ValueTuple`2 data)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.ExecuteAndCatchIfThrows_NoLock[TArg](DiagnosticAnalyzer analyzer, Action`1 analyze, TArg argument, Nullable`1 info)
-----

Suppress the following diagnostics to disable this analyzer: DF0101, DF0102, DF0103, DF0104, DF0105, DF0106, DF0107, DF0111	<project name>		1	Active

```

### Known workarounds
None

### App Details

- **Durable Functions extension version (e.g. v1.8.3)**: 2.4.3
- **Azure Functions runtime version (1.0 or 2.0)**: 3
- **Programming language used**: .NET
- The following in project file:
  - `<EnableNETAnalyzers>true</EnableNETAnalyzers>`
  - `<AnalysisMode>AllEnabledByDefault</AnalysisMode>`
  - `<TreatWarningsAsErrors>true</TreatWarningsAsErrors>`
  - `<WarningsNotAsErrors>612,618</WarningsNotAsErrors>`
   - `<NoWarn>1701;1702;1591;VSTHRD200</NoWarn>`

### PR DESCRIPTION
**Pull request by amdeel**
Analyzer updates for 0.4.1 release:
- Incremented version to 0.4.1
- Updated Entity InterfaceContentAnalyzer to avoid flagging interfaces that implement another interface.
  - resolves #1704 
  - _InterfaceContentAnalyzer.cs, InterfaceContentAnalyzerTests.cs_

- Added BindingAnalyzer to deterministic method checks, flagging all attributes used on parameters that are not OrchestrationTrigger to prevent customers from using additional bindings on their orchestrator method.
  - resolves #1879
  - _BindingAnalyzer.cs, BindingAnalyzerTests.cs_

- Fixed ArgumentException thrown in SyntaxUtils.TryGetSemanticModelForSyntaxTree that occurred when the SyntaxNode being analyzed was not in the current compilation by creating a new compilation that adds the syntax tree for the syntax node. Also added a catch block in TryGetSemanticModelForSyntaxTree for language version mismatches.
  - resolves #1883 
  - _SyntaxNodeUtils.cs, OrchestratorMethodCollector.cs_

### Pull request checklist

* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation PR is ready to merge and referenced in `pending_docs.md`
* [x] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] I have added all required tests (Unit tests, E2E tests)
* [x] My changes **do not** require any extra work to be leveraged by OutOfProc SDKs
    * [ ] Otherwise: That work is being tracked here: #issue_or_pr_in_each_sdk

### CODE REVIEW

File: src/WebJobs.Extensions.DurableTask.Analyzers/SyntaxNodeUtils.cs [hunk 1]
```diff
@@ -29,16 +29,34 @@ public static DurableVersion GetDurableVersion(SemanticModel semanticModel)
...
+                else
+                {
+                    newModel = model.SyntaxTree == node.SyntaxTree
+                    ? model
+                    : model.Compilation.GetSemanticModel(node.SyntaxTree);
+                }
+
+                return newModel != null;
+            }
+            catch( ArgumentException e) when (e.Message == "Inconsistent language versions")
```
Comment (amdeel) [inline src/WebJobs.Extensions.DurableTask.Analyzers/SyntaxNodeUtils.cs:55]: I couldn't find a good way to check this so I had to catch this exception.

Comment (ConnorMcMahon) [reply src/WebJobs.Extensions.DurableTask.Analyzers/SyntaxNodeUtils.cs:55]: When do we expect to hit this exception? a comment describing when we expect to hit this control flow would be super helpful.

File: src/WebJobs.Extensions.DurableTask.Analyzers/SyntaxNodeUtils.cs [hunk 2]
```diff
@@ -29,16 +29,34 @@ public static DurableVersion GetDurableVersion(SemanticModel semanticModel)
 
         public static bool TryGetSemanticModelForSyntaxTree(SemanticModel model, SyntaxNode node, out SemanticModel newModel)
         {
-            if (model?.SyntaxTree == null || node?.SyntaxTree == null)
+            try
+            {
+                if (model?.SyntaxTree == null || node?.SyntaxTree == null)
```
Comment (ConnorMcMahon) [inline src/WebJobs.Extensions.DurableTask.Analyzers/SyntaxNodeUtils.cs:34]: nit: this try catch can be moved outside the try-catch.

In general, if you have identified which method can throw the ArgumentException, I would only wrap that code in the try-catch to make it more clear where we expect the exception to be thrown.

File: src/WebJobs.Extensions.DurableTask.Analyzers/SyntaxNodeUtils.cs [hunk 3]
```diff
@@ -29,16 +29,34 @@ public static DurableVersion GetDurableVersion(SemanticModel semanticModel)
...
+                    newModel = newComplilation.GetSemanticModel(node.SyntaxTree);
+                }
+                else
+                {
+                    newModel = model.SyntaxTree == node.SyntaxTree
+                    ? model
+                    : model.Compilation.GetSemanticModel(node.SyntaxTree);
+                }
+
+                return newModel != null;
```
Comment (ConnorMcMahon) [inline src/WebJobs.Extensions.DurableTask.Analyzers/SyntaxNodeUtils.cs:53]: Is it even possible for newModel to be null at this point?

**Review by ConnorMcMahon — COMMENTED** [review]
Mostly nit picks and a few questions. I think this is pretty close to merging!

**Pull request merged**